File: Pharmaceutical_Company_Example/02_Structural/R/Decorator.R

# Base class
RegulatoryProcess <- setRefClass(
  "RegulatoryProcess",
  methods = list(
    execute = function() {
      stop("This method should be implemented by subclasses.")
    }
  )
)

# Concrete class
BasicRegulatoryStep <- setRefClass(
  "BasicRegulatoryStep",
  contains = "RegulatoryProcess",
  fields = list(step_name = "character"),
  methods = list(
    initialize = function(step_name) {
      .self$step_name <- step_name
    },
    execute = function() {
      message("Executing step: ", .self$step_name)
    }
  )
)

# Decorator Base Class for adding additional features to RegulatoryProcess
ProcessDecorator <- setRefClass(
  "ProcessDecorator",
  contains = "RegulatoryProcess",
  fields = list(decorated_process = "ANY"),
  methods = list(
    initialize = function(decorated_process) {
      .self$decorated_process <- decorated_process
    },
    execute = function() {
      .self$decorated_process$execute()
    }
  )
)

# Concrete Decorator - Adds Auditing feature to the regulatory process
AuditingDecorator <- setRefClass(
  "AuditingDecorator",
  contains = "ProcessDecorator",
  methods = list(
    execute = function() {
      .self$decorated_process$execute()
      .self$add_audit()
    },
    add_audit = function() {
      message("Adding auditing to step.")
    }
  )
)

# Concrete Decorator - Adds Documentation requirement to the process
DocumentationDecorator <- setRefClass(
  "DocumentationDecorator",
  contains = "ProcessDecorator",
  methods = list(
    execute = function() {
      .self$decorated_process$execute()
      .self$add_documentation()
    },
    add_documentation = function() {
      message("Ensuring documentation is complete for step.")
    }
  )
)

clinical_trial_step <- BasicRegulatoryStep$new("Clinical Trial Approval")
safety_review_step <- BasicRegulatoryStep$new("Safety Review")

# Decorate the clinical trial step with auditing and documentation features
clinical_trial_with_audit <- AuditingDecorator$new(clinical_trial_step)
clinical_trial_with_full_features <- DocumentationDecorator$new(clinical_trial_with_audit)

# Decorate the safety review step with documentation only
safety_review_with_documentation <- DocumentationDecorator$new(safety_review_step)

# Execute both processes
message("Executing Clinical Trial Step with Audit and Documentation:")
clinical_trial_with_full_features$execute()

message("\nExecuting Safety Review Step with Documentation:")
safety_review_with_documentation$execute()

File: Pharmaceutical_Company_Example/02_Structural/PHP/Decorator.php
<?php

// Base class: RegulatoryProcess
abstract class RegulatoryProcess {
    abstract public function execute(): void;
}

// Concrete class: BasicRegulatoryStep
class BasicRegulatoryStep extends RegulatoryProcess {
    private string $stepName;

    public function __construct(string $stepName) {
        $this->stepName = $stepName;
    }

    public function execute(): void {
        echo "Executing step: {$this->stepName}\n";
    }
}

// Decorator Base Class
abstract class ProcessDecorator extends RegulatoryProcess {
    protected RegulatoryProcess $decoratedProcess;

    public function __construct(RegulatoryProcess $decoratedProcess) {
        $this->decoratedProcess = $decoratedProcess;
    }

    public function execute(): void {
        $this->decoratedProcess->execute();
    }
}

// Concrete Decorator - Adds Auditing Feature
class AuditingDecorator extends ProcessDecorator {
    public function execute(): void {
        parent::execute();
        $this->addAudit();
    }

    private function addAudit(): void {
        echo "Adding auditing to step.\n";
    }
}

// Concrete Decorator - Adds Documentation Requirement
class DocumentationDecorator extends ProcessDecorator {
    public function execute(): void {
        parent::execute();
        $this->addDocumentation();
    }

    private function addDocumentation(): void {
        echo "Ensuring documentation is complete for step.\n";
    }
}

// Concrete steps
$clinicalTrialStep = new BasicRegulatoryStep("Clinical Trial Approval");
$safetyReviewStep = new BasicRegulatoryStep("Safety Review");

// Decorate the clinical trial step with auditing and documentation features
$clinicalTrialWithAudit = new AuditingDecorator($clinicalTrialStep);
$clinicalTrialWithFullFeatures = new DocumentationDecorator($clinicalTrialWithAudit);

// Decorate the safety review step with documentation only
$safetyReviewWithDocumentation = new DocumentationDecorator($safetyReviewStep);

// Execute both processes
echo "Executing Clinical Trial Step with Audit and Documentation:\n";
$clinicalTrialWithFullFeatures->execute();

echo "\nExecuting Safety Review Step with Documentation:\n";
$safetyReviewWithDocumentation->execute();

?>
